src/lobpcg.jl

```diff
@@ -0,0 +1,751 @@
+    M::TM
+    buffer::TA
+    RPreconditioner{TM, T, TA}(M, X) where {TM, T, TA<:AbstractArray{T}} = new(M, similar(X))
+end
+RPreconditioner(M, X) = RPreconditioner{typeof(M), eltype(X), typeof(X)}(M, X)
+
+function (precond!::RPreconditioner{Void})(X)
+    nothing
+end
+function (precond!::RPreconditioner)(X)
+    bs = size(X, 2)
+    A_ldiv_B!(view(precond!.buffer, :, 1:bs), precond!.M, X)
+    # Just returning buffer would be cheaper but struct at call site must be mutable
+    X .= view(precond!.buffer, :, 1:bs)
+    nothing
+end
+
+struct BlockGram{Generalized, TA}
+    XAX::TA
+    XAR::TA
+    XAP::TA
+    RAR::TA
+    RAP::TA
+    PAP::TA
+end
+function BlockGram(XBlocks::Blocks{Generalized, T}) where {Generalized, T}
+    sizeX = size(XBlocks.block, 2)
+    XAX = zeros(T, sizeX, sizeX)
+    XAP = zeros(T, sizeX, sizeX)
+    XAR = zeros(T, sizeX, sizeX)
+    RAR = zeros(T, sizeX, sizeX)
+    RAP = zeros(T, sizeX, sizeX)
+    PAP = zeros(T, sizeX, sizeX)
+    return BlockGram{Generalized, Matrix{T}}(XAX, XAR, XAP, RAR, RAP, PAP)
+end
+XAX!(BlockGram, XBlocks) = Ac_mul_B!(BlockGram.XAX, XBlocks.block, XBlocks.A_block)
+XAP!(BlockGram, XBlocks, PBlocks, n) = Ac_mul_B!(view(BlockGram.XAP, :, 1:n), XBlocks.block, view(PBlocks.A_block, :, 1:n))
+XAR!(BlockGram, XBlocks, RBlocks, n) = Ac_mul_B!(view(BlockGram.XAR, :, 1:n), XBlocks.block, view(RBlocks.A_block, :, 1:n))
+RAR!(BlockGram, RBlocks, n) = Ac_mul_B!(view(BlockGram.RAR, 1:n, 1:n), view(RBlocks.block, :, 1:n), view(RBlocks.A_block, :, 1:n))
+RAP!(BlockGram, RBlocks, PBlocks, n) = Ac_mul_B!(view(BlockGram.RAP, 1:n, 1:n), view(RBlocks.A_block, :, 1:n), view(PBlocks.block, :, 1:n))
+PAP!(BlockGram, PBlocks, n) = Ac_mul_B!(view(BlockGram.PAP, 1:n, 1:n), view(PBlocks.block, :, 1:n), view(PBlocks.A_block, :, 1:n))
+XBP!(BlockGram, XBlocks, PBlocks, n) = Ac_mul_B!(view(BlockGram.XAP, :, 1:n), XBlocks.block, view(PBlocks.B_block, :, 1:n))
+XBR!(BlockGram, XBlocks, RBlocks, n) = Ac_mul_B!(view(BlockGram.XAR, :, 1:n), XBlocks.block, view(RBlocks.B_block, :, 1:n))
+RBP!(BlockGram, RBlocks, PBlocks, n) = Ac_mul_B!(view(BlockGram.RAP, 1:n, 1:n), view(RBlocks.B_block, :, 1:n), view(PBlocks.block, :, 1:n))
+XBX!(BlockGram, XBlocks) = Ac_mul_B!(BlockGram.XAX, XBlocks.block, XBlocks.B_block)
+RBR!(BlockGram, RBlocks, n) = Ac_mul_B!(view(BlockGram.RAR, 1:n, 1:n), view(RBlocks.block, :, 1:n), view(RBlocks.B_block, :, 1:n))
+PBP!(BlockGram, PBlocks, n) = Ac_mul_B!(view(BlockGram.PAP, 1:n, 1:n), view(PBlocks.block, :, 1:n), view(PBlocks.B_block, :, 1:n))
+
+function I!(G, xr)
+    for j in xr, i in xr
+        G[i, j] = ifelse(i==j, 1, 0)
+    end
+    return
+end
+
+function (g::BlockGram)(gram, lambda, n1::Int, n2::Int, n3::Int)
+    xr = 1:n1
+    rr = n1+1:n1+n2
+    pr = n1+n2+1:n1+n2+n3
+    if n1 > 0
+        #gram[xr, xr] .= view(g.XAX, 1:n1, 1:n1)
+        gram[xr, xr] .= Diagonal(view(lambda, 1:n1))
+    end
+    if n2 > 0
+        gram[rr, rr] .= view(g.RAR, 1:n2, 1:n2)
```

Comment:
When dotting you don't need views for the rhs, indexing is wrapped in `dotview`. Just putting the comment here, but applies everywhere.

Reply:
Is this in v0.6?
```julia
julia> A = zeros(300000);

julia> @time A[1:end] .= A[1:end];
  0.003292 seconds (332 allocations: 2.308 MiB)

julia> @time A[1:end] .= A[1:end];
  0.001019 seconds (32 allocations: 2.290 MiB)

julia> @time A[1:end] .= view(A, 1:length(A));
  0.000241 seconds (47 allocations: 1.406 KiB)

julia> @time A[1:end] .= view(A, 1:length(A));
  0.000181 seconds (46 allocations: 1.359 KiB)
```

Reply:
Ah, I was confused! Only the lhs is wrapped in dotview. For some heavy `view` sections you could also use the `@views` macro which wraps indexing ops in `maybeview`, which might be slightly cleaner -- but that's for you to judge.

Reply:
I will stick with the current one then. It's less appealing but more informative about what's happening.